timdex_dataset_api/dataset.py

```diff
@@ -111,23 +133,127 @@ def load_dataset(self) -> ds.Dataset:
         raised when reading or writing data.
         """
         start_time = time.perf_counter()
-        filesystem, source = self.parse_location(self.location)
         dataset = ds.dataset(
-            source,
+            self.source,
             schema=self.schema,
             format="parquet",
             partitioning="hive",
-            filesystem=filesystem,
+            filesystem=self.filesystem,
         )
         logger.info(
             f"Dataset successfully loaded: '{self.location}', "
             f"{round(time.perf_counter()-start_time, 2)}s"
         )
         return dataset
 
+    def reload(self) -> None:
+        """Reload dataset.
+
+        After a write has been performed, a reload of the dataset is required to reflect
+        any newly added records.
+        """
+        self.dataset = self.load_dataset()
+
     @property
     def row_count(self) -> int:
         """Get row count from loaded dataset."""
         if not self.dataset:
             raise DatasetNotLoadedError
         return self.dataset.count_rows()
+
+    def write(
+        self,
+        records_iter: Iterator["DatasetRecord"],
+        *,
+        partition_values: dict[str, str | datetime.datetime] | None = None,
+        batch_size: int = DEFAULT_BATCH_SIZE,
+        use_threads: bool = True,
+    ) -> list[ds.WrittenFile]:
+        """Write records to the TIMDEX parquet dataset.
+
+        This method expects an iterator of DatasetRecord instances, with optional
+        partition column values that will be applied to all rows written (often, these
+        are the same for all rows written, eliminating the need to repeat those values
+        in the iterator).
+
+        This method encapsulates all dataset writing mechanics and performance
+        optimizations (e.g. batching) so that the calling context can focus on yielding
+        data.
+
+        Args:
+            - records_iter: Iterator of DatasetRecord instances
+            - partition_values: dictionary of static partition column name/value pairs
+            - batch_size: size for batches to yield and write, directly affecting row
+                group size in final parquet files
+            - use_threads: boolean if threads should be used for writing
+        """
+        start_time = time.perf_counter()
+
+        if isinstance(self.source, list):
+            raise TypeError(
+                "Dataset location must be the root of a single dataset for writing"
+            )
+
+        record_batches_iter = self.get_dataset_record_batches(
+            records_iter,
+            partition_values=partition_values,
+            batch_size=batch_size,
+        )
+
+        written_files = []
+        ds.write_dataset(
+            record_batches_iter,
+            base_dir=self.source,
+            basename_template="%s-{i}.parquet" % (str(uuid.uuid4())),  # noqa: UP031
+            existing_data_behavior="delete_matching",
```

Comment:
Why does `existing_data_behavior="delete_matching"`? I see that there are [tests in `test_dataset_write.py` for some of the different configurations](https://github.com/MITLibraries/timdex-dataset-api/pull/9/files#diff-84a497f3c25d904abbb5dcbdc3bd0e195a3713f1e56a558099ca6285a6e6ffcaR86-R181). Can you add one for `existing_data_behavior` as well? 🤔 

In what case(s) would there be "existing data" that would get overwritten? Is this only in the case of reruns?

Reply:
Good idea!  It's possible that we could get away with `existing_data_behavior="overwrite_or_ignore"`, given that we are minting UUIDs for each file written.  In that scenario, even if we wrote to the same exact partition location, a new file would always be created.

With `existing_data_behavior="delete_matching"` we get this behavior from pyarrow:

> ‘delete_matching’ is useful when you are writing a partitioned dataset. The first time each partition directory is encountered the entire directory will be deleted. This allows you to overwrite old partitions completely.

In this way, if we were to ever write to the exact same partition, it would actually delete anything under that partition  combination.  Because partition `run_id` is supposed to be unique for each TIMDEX "run" (i.e. time when dataset is written to), it feels correct that we'd want to blow away anything under that `run_id` partition and let this new version take precedence.   If I'm thinking about this correctly, I think this would make runs mostly **idempotent**.   Even if we somehow re-ran the StepFunction with the same execution ID it would fully remove and recreate everything under that `run_id` partition.

But you're right, this is ripe for a test or two!  Adding.